pkg/api/endpoints/util.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoints
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"hash"
+	"sort"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// RepackSubsets takes a slice of EndpointSubset objects, expands it to the full
+// representation, and then repacks that into the canonical layout.  This
+// ensures that code which operates on these objects can rely on the common
+// form for things like comparison.  The result is a newly allocated slice.
+func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
+	// First map each unique port definition to the sets of hosts that
+	// offer it.  The sets of hosts must be de-duped, using IP as the key.
+	type ipString string
+	allAddrs := map[ipString]*api.EndpointAddress{}
+	portsToAddrs := map[api.EndpointPort]addressSet{}
+	for i := range subsets {
+		for j := range subsets[i].Ports {
+			epp := &subsets[i].Ports[j]
+			for k := range subsets[i].Addresses {
```

Comment:
Looping over values causes copies.  Basically, any time there's a map or
slice where the value is anything but a primitive type you incur totally
pointless copies.  Go just encourages you to piss away your performance.

I asked them about this once and they simply said, "yeah, just iterate
slices in the C style, with a pointer to the current element".  So that's
the pattern I follow everywhere except test code where it doesn't matter.

In short `map[anything]struct` and `for x, y := range slice` are traps.

On Thu, Mar 26, 2015 at 9:35 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/api/endpoints/util.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5939#discussion_r27275581>
> :
>
> > +)
> > +
> > +// RepackSubsets takes a slice of EndpointSubset objects, expands it to the full
> > +// representation, and then repacks that into the canonical layout.  This
> > +// ensures that code which operates on these objects can rely on the common
> > +// form for things like comparison.  The result is a newly allocated slice.
> > +func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
> > +	// First map each unique port definition to the sets of hosts that
> > +	// offer it.  The sets of hosts must be de-duped, using IP as the key.
> > +	type ipString string
> > +	allAddrs := map[ipString]*api.EndpointAddress{}
> > +	portsToAddrs := map[api.EndpointPort]addressSet{}
> > +	for i := range subsets {
> > +		for j := range subsets[i].Ports {
> > +			epp := &subsets[i].Ports[j]
> > +			for k := range subsets[i].Addresses {
>
> Go question: Why not loop over the values instead of the keys?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5939/files#r27275581>
> .
>